### DISCUSSION TITLE
Two new node suggestions - SWITCH and GATE

### DISCUSSION THREAD

**Discussion by Vantiveman** [comfyanonymous/ComfyUI#1625] [Ideas]
Complex node configurations can include much functionality that can be enabled or disabled, depending on what is being done. Currently this requires that "noodles" are disconnected and later connected again when the extra functionality is required, like adding LORA or refiner etc. This is often complicated as noodles can be extensive and working out where to reconnect can be a bit of a pain.
I would like to suggest the following Nodes.
1) SWITCH: This node has a single input with one or more outputs and a single parameter, STATE, which would be a boolean. If the State is true then the switch is enabled and the input passes through to the output(s). If False then the switch turns off the passthrough of the input. This would remove the need to disconnect noodles and remember where they go to re-instate later. One to many allows more clarify of where the noodles connect as they would not all originate from the same point.
Is this possible? 
2) GATE: This node would act like a logic gate and have 2 or more inputs and one output. The output would create the STATE input to the SWITCH node (if converted to an input). It would function as follows with 4 variants:
a) AND. Where all inputs have a connected noodle then the output state is TRUE.
b) OR. Where any of the inputs have a connected noodle then the output is TRUE.
c) NAND. The inverse of AND with the output being True only when all inputs are disconnected.
d) NOR. The inverse of OR where the output would be true only when any input is disconnected.
(I hope I have the logic correct)
I think that, if possible, this would enable the SWITCH node operation to be automated based on the GATE logic, enabling dynamic operation where sections of the node configuration are enabled or disabled depending on SWITCH states or the state of the GATEs which could turn the switches on or off.
As I know nothing about the design of nodes I really have no idea whether the above suggestion is possible to implement without nodes with inputs turned off turning red and causing the prompt to fail. However, having done a lot of work on electronics and TTL circuitry I thought the similarity of the node configuration to an electronic circuit was an opportunity too good to miss.
Apologies if this idea makes no sense or is simply impossible. It's an idea I couldn't get out of my head!

**Comment by ghostsquad** [comment]
Came here looking for this. E.g. Some models have a built-in VAE, but I'd like to be able to sometimes use a separate VAE. SWITCH would work well for that scenario.

**Reply by el0911** [reply]
try https://github.com/virgincodes/comfyIfElse-node

**Comment by ghostsquad** [comment]
I have a PR to WAS to add some logic nodes (AND, OR, XOR, NOT)

https://github.com/WASasquatch/was-node-suite-comfyui/pull/331

and Impact has a node called ImpactConditionalBranch: https://github.com/ltdrdata/ComfyUI-Impact-Pack

**Comment by el0911** [comment]
<img width="855" alt="Screen Shot 2024-01-18 at 1 02 30" src="https://github.com/comfyanonymous/ComfyUI/assets/11575902/760f3218-bcea-42f7-8ad5-72fe00a79aed">
@ghostsquad  this is what am  trying to acheive

**Comment by el0911** [comment]
to conditionally pass 1 image to only one node among the two and the other one just remains frozen

**Reply by ghostsquad** [reply]
I'm current doing this with rgthree "any switch" and "fast muter". It's a good pattern for manual decisions (switching between inputs). It seems what you want to do is change the "flow of the river" if you will... a single input can go left or right. 

My research into ComfyUI is that this type of logic isn't entirely possible at runtime. Possible at queue time though, as you have built. 

The reason why is because ComfyUI builds the workflow _backwards_ starting at leaf nodes and working backwards to identify what dependent nodes need to run. This is also why it's fast, and how it can cache work that hasn't changed.

**Reply by el0911** [reply]
well that is right i am trying to change the flow of the ocean , but is their any node that completely switches of the subsequent nodes after it?

**Reply by el0911** [reply]
lol i just got it to work i took ur advice and edited my comfyui source code yeah works now with the example dont know if it work well on a big scale would check though

**Reply by ghostsquad** [reply]
Not automatically, since workflows are "built" in reverse, then executed based on the graph built from that. 

You can control the forward flow manually, by muting/bypassing,  before queuing because pre-queued decisions affect how the graph is built.

**Reply by el0911** [reply]
what u mean by muting? or bypassing

**Reply by ghostsquad** [reply]
A muted node is one that doesn't run. Muting though must be done before the graph is built (queued). For a bypassed node, Comfy essentially tries to connect the inputs directly to the outputs, as if that node doesn't exist. Those are "modes" of a node, that you can quickly change before queuing a workflow. Example, select a node and hit "m" on your keyboard. The node won't run, in that way though, it also "blocks" downstream nodes from running too, since those nodes don't have their dependencies met.

**Reply by el0911** [reply]
oh i get but muting a node and unmuting it the graph gets built every time any of those happen?

**Reply by ghostsquad** [reply]
Muting is something _you_ do, not something that can be done automatically (at runtime)

**Reply by ghostsquad** [reply]
Damn autocorrect

**Reply by el0911** [reply]
i assume that the information of a muted node would be in the schema sent fro the frontend to the backend ?

**Reply by ghostsquad** [reply]
That's correct

**Reply by el0911** [reply]
ok i get but funny my solution does work at the moment and actually changes the flow of data it seams all the nodes are built it doesnt just pass data to the nodes that it isnt told to effectively making a contional noder

**Reply by ghostsquad** [reply]
The feature that you and I both want, is essentially sub/dynamic workflows, which would combine the many of the benefits of a "compiled" workflow, with that of dynamic decisions. Essentially, the first workflow/graph runs, producing some set of outputs. Those outputs are plumbed into the next graph, allowing that one to be compiled and cached in-turn based on input data that has/hasn't changed.

**Reply by ghostsquad** [reply]
Some nodes may error if they don't get the data they expect, so the problem is that things still end up running and possibly erroring.

**Reply by el0911** [reply]
yeah , i want sub workflows so i can at one single runtime compile multiple workflows and run them on one comfy server instead of creating multiple comfyservers , or rebuilding a new workflow i can now run all of them on one server and just tell it which pipleine i want it to run

**Reply by ghostsquad** [reply]
I just realized, this type of thing may be possible with a ComfyUI wrapper. Basically, anything that interacts directly the with API of comfyUI could execute a workflow, get some values from that execution, populate another workflow with those values, then queue that up. 

This type of feature definitely seems very doable natively I think from the current UI.

I think the only thing really missing is a concept of a workflow output, though that could be done by just writing a file to disk. The second would be a trigger mechanism. The final bit would be a way to input the values read into the next workflow before submitting it to the API.

**Comment by davslye** [comment]
I've been experimenting with a similar concept to this, with the nodes named

## Winnow:
* With multiple input options, select only one output (of the connected inputs)
![image](https://github.com/comfyanonymous/ComfyUI/assets/14813056/9db38029-868d-402d-8156-02dfc6faf60d)


## Fork:
* With exactly one input option, redirect to one of the output options

**Reply by benedlore** [reply]
This is exactly what I need. Ever complete this?

**Reply by ltdrdata** [reply]
FYI, Impact Pack supports these features.
https://www.youtube.com/watch?v=KLe1hShwVDs

**Reply by benedlore** [reply]
Hey thanks! Those work well.
